### PR TITLE
Pull Request for peer review

### PR DESCRIPTION
**Pull request by anuonifade**
#### What does this PR do?
- Adds files on development to master branch

#### Description of Task to be completed?
- Sets up travis ci on the repository
- Sets up coveralls 
- Sets up hound
- Sets up eslint
- Adds inverted index class files.
- Adds inverted index controller
- Adds inverted index html files
- Adds inverted index css files
- Adds inverted index javascript and css dependencies

#### How should this be manually tested?
- NA
#### Any background context you want to provide?
- NA
#### What are the relevant pivotal tracker stories?
- NA
#### Screenshots (if appropriate)
- NA
#### Questions:
- NA

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+gulp.task('serve', ['watch'], () => {
+  browserSync.init({
+    server: {
+      baseDir: './'
+    }
+  });
+});
+
+gulp.task('lint', () => {
+  // ESLint ignores files with "node_modules" paths.
```
Comment (raphaelakpan) [inline gulpfile.js]: Why do you have this much lines of comments on this task?

File: spec/dplex-test-spec.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      expect(typeof this.indexInstance.getSearchResults).toBe('function');
+    });
+
+    it('should check that the class has a getDocuments method', () => {
+      expect(typeof this.indexInstance.getDocuments).toBe('function');
+    });
+  });
+
+  describe('Create Index', () => {
+    it('should return mapped indices to words in a JSON file', () => {
```
Comment (raphaelakpan) [inline spec/dplex-test-spec.js:62]: Did you test for when the book was not successfully created? Or do you expect `createIndex` to always succeed?

File: spec/dplex-test-spec.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+          lord: [1, 2],
+          rings: [1, 2],
+          fellowship: [1, 2] };
+      this.indexInstance.createIndex(this.books, 'books');
+      expect(this.indexInstance.filesIndexed.books.index)
+        .toEqual(expectedResult);
+    });
+  });
+  describe('Search Index', () => {
+    it('should search through single files that are indexed', () => {
```
Comment (raphaelakpan) [inline spec/dplex-test-spec.js]: What if the search is not successful?

File: spec/dplex-test-spec.js [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  describe('Tokenize words', () => {
+    it('should strip out special characters from excerpt in documents', () => {
+      let excerpt = 'Alice l##$oves her ima&&gination?';
+      const expectedTokens = ['alice', 'loves', 'her', 'imagination'];
+      excerpt = this.indexInstance.tokenize(excerpt);
+      expect(excerpt).toEqual(expectedTokens);
+    });
+  });
+
+  describe('Search index', () => {
```
Comment (raphaelakpan) [inline spec/dplex-test-spec.js]: You already declared a `describe` block for `Search Index`. Put all `searchIndex` related tests inside that `describe` block to easy of readability.

File: spec/dplex-test-spec.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        lord: [1, 2],
+        rings: [1, 2],
+        fellowship: [1, 2] };
+      this.indexInstance.createIndex(this.books, 'books');
+      expect(this.indexInstance.getIndex(filename))
+        .toEqual(expectedOutput);
+    });
+  });
+
+  describe('Get index', () => {
```
Comment (raphaelakpan) [inline spec/dplex-test-spec.js]: Use one `describe` to logically group your tests

File: src/app.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
+
+dPlexApp.controller('dPlexController', ['$scope', 'toastr', ($scope, toastr) => {
+  $scope.title = 'DPlex - Inverted Index for All';
+  // Create an object of the class InvertedIndex
+  const invIndex = new InvertedIndex();
```
Comment (raphaelakpan) [inline src/app.js]: Why not use complete and descriptive variable name?

File: src/invertedindex.js [hunk 7]
```diff
@@ -0,0 +1,222 @@
+
+const hasProperty = Object.prototype.hasOwnProperty;
+let instance = null;
+
+class InvertedIndex {
+  /**
+    * constructor method
+  */
+  constructor() {
+    if (!instance) {
```
Comment (femisenjobi) [inline src/invertedindex.js]: The constructor does not take any parameters. This block will always be hit. I thnk it makes sense to take it out of the conditional.

File: src/invertedindex.js [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    });
+  }
+
+  /**
+   * prepareJsonData gets the json ready for indexing by tokenizing statements
+   * @param {object} inputData - the json data to index
+   * @param {string} filename - the name of the file to be indexed
+   * @return {boolean} true or false if the createIndex was successful
+  */
+  createIndex(inputData, filename) {
```
Comment (femisenjobi) [inline src/invertedindex.js]: It is always good practice to keep methods as short and as specific as possible. You can extract line 56-59 into a function and reuse at line 63-66. The function may look like this.
```
handleError(fileName, errorMessage, ) {
    delete this.filesIndexed[fileName];
    this.error.status = true;
    this.error.message = errorMessage;
    throw this.error;
}
```
You can now have something like
```
try {
     if (Object.keys(inputData).length < 1) {
          this.handleError(filename, 'File contains no document');
     }
     Object.keys(inputData).forEach((eachIndex) => {
          if (!this.validateFile(inputData[eachIndex])) {
              this.handleError(filename, 'Incorrect Document Structure');
         }
         words.push(this.getDocumentTokens(inputData, documentNum));
         documentNum += 1;
     });
```
This appears neater.

File: src/invertedindex.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+  /**
+   * getIndex method returns the indexed words and the documents that were found
+   * @param filename {string} : name of the file to get its index
+   * @return {Object} or {boolean} the words index
+  */
+  getIndex(filename) {
+    try {
+      if (!this.filesIndexed[filename]) {
+        this.error.status = false;
```
Comment (femisenjobi) [inline src/invertedindex.js]: You can reuse handleError here too

File: src/invertedindex.js [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+  }
+
+  /**
+   * getDocuments get an array of the documents index e.g [0, 1, 2, 3]
+   * @param {filename} - name of the file to get its document
+   * @return {array} an array of the documents index
+  */
+  getDocuments(filename) {
+    const docs = [];
+    for (let i = 0; i < this.filesIndexed[filename].numOfDocs; i += 1) {
```
Comment (femisenjobi) [inline src/invertedindex.js]: `i++` works preety fine too.

File: gulpfile.js (hunk 1 above)
Comment (anuonifade) [reply gulpfile.js]: Removed.

File: spec/dplex-test-spec.js (hunk 2 above)
Comment (anuonifade) [reply spec/dplex-test-spec.js:62]: Done! It was initially separated but it's been rectified to be together.

File: spec/dplex-test-spec.js (hunk 4 above)
Comment (anuonifade) [reply spec/dplex-test-spec.js]: corrected

File: spec/dplex-test-spec.js (hunk 5 above)
Comment (anuonifade) [reply spec/dplex-test-spec.js]: done. thanks

File: src/app.js (hunk 6 above)
Comment (anuonifade) [reply src/app.js]: done.Thanks

File: src/invertedindex.js (hunk 7 above)
Comment (anuonifade) [reply src/invertedindex.js]: The idea is to create a singleton class where object are created just once. I tested it and it worked perfectly well. I hope that makes sense except maybe it's against common practice. Please advise.

File: src/invertedindex.js (hunk 9 above)
Comment (anuonifade) [reply src/invertedindex.js]: Done. Thanks

File: src/invertedindex.js (hunk 10 above)
Comment (anuonifade) [reply src/invertedindex.js]: Done. Thanks

File: src/invertedindex.js (hunk 8 above)
Comment (anuonifade) [reply src/invertedindex.js]: Done thanks

**Pull request merged**